### PR TITLE
Fix unauthorized access to chain balance

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

It's currently possible for an application to transfer funds from the receiving chain's shared owner balance when receiving a message. The only authentication is off-chain, where it is assumed that owners will reject messages that move the shared chain balance without permission. However, by default the node services automatically receives messages. This means that it's too easy to mistakenly accept a message that transfers the shared chain balance tokens without the appropriate permission.

## Proposal

Reject transfer attempts when the authenticated signer is not one of the chain owners.

## Test Plan

A new unit test was added to cover the scenario. The test fails before the fix and passes after the fix.

## Release Plan

- These changes should be backported to the latest `devnet` branch and `testnet` branch, because it improves security of chain accounts.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-execution/src/system.rs [hunk 1]
```diff
@@ -668,11 +668,16 @@ where
...
-        if owner.is_some() {
-            ensure!(
+        match (owner, authenticated_signer) {
+            (Some(_), _) => ensure!(
                 authenticated_signer == owner,
                 SystemExecutionError::UnauthenticatedTransferOwner
-            );
+            ),
+            (None, Some(signer)) => ensure!(
+                self.ownership.get().verify_owner(&signer).is_some(),
```
Comment (ma2bd) [inline linera-execution/src/system.rs:677]: Good catch

**Review by ma2bd — APPROVED** [review]

**Review by afck — APPROVED** [review]

File: linera-execution/tests/test_execution.rs [hunk 2]
```diff
@@ -1667,3 +1660,90 @@ async fn test_close_chain() {
...
+
+    let mut state = SystemExecutionState::default();
+    state.description = Some(ChainDescription::Root(0));
+    state.balance = amount;
+    let mut view = state.into_view().await;
+
+    let mut applications = register_mock_applications(&mut view, 1).await?;
+    let (application_id, application, _, _) = applications
+        .next()
+        .expect("Caller mock application should be registered");
```
Comment (deuszx) [inline linera-execution/tests/test_execution.rs:1678]: Alternatively a version of `register_mock_application` which returns the application - no need to `.next().expect()`.

File: linera-execution/tests/test_execution.rs [hunk 3]
```diff
@@ -1667,3 +1660,90 @@ async fn test_close_chain() {
     .unwrap();
     assert!(view.system.closed.get());
 }
+
+/// Tests if an application can't transfer the tokens in the chain's balance while executing
+/// messages.
+#[tokio::test]
+async fn test_message_handler_cant_spend_chain_balance() -> anyhow::Result<()> {
```
Comment (deuszx) [inline linera-execution/tests/test_execution.rs:1667]: Do we have a test for the "happy path"? Also, it looks like you've added two match cases that fail with unauthenticated owner errors but we have a test for one, right?

**Comment by ma2bd** [comment]
re: release plan: I don't think this is possible because (in theory) it could make certified blocks potentially invalid.

**Comment by jvff** [comment]
> @jvff re: release plan: I don't think this is possible because (in theory) it could make some certified blocks invalid.

That's true, but I was assuming that this edge case had not been triggered yet :thinking: I'm fine with leaving it for the next version, though.

File: linera-execution/tests/test_execution.rs (hunk 2 above)
Comment (jvff) [reply linera-execution/tests/test_execution.rs]: I agree, I opened issue #2851 to track this.

File: linera-execution/tests/test_execution.rs (hunk 3 above)
Comment (jvff) [reply linera-execution/tests/test_execution.rs]: I added some more test cases, but they still don't cover the first `match` arm, which is equivalent to the previous `if` condition, because that's out of this PR's scope (access to the balance shared by all chain owners). I opened issue #1852 to track that though.

**Comment by ma2bd** [comment]
> > @jvff re: release plan: I don't think this is possible because (in theory) it could make some certified blocks invalid.
> 
> That's true, but I was assuming that this edge case had not been triggered yet 🤔 I'm fine with leaving it for the next version, though.

e2e tests might have triggered it. We probably don't want to find out in prod.
